pyteal/ast/subroutine.py

```diff
@@ -519,16 +519,27 @@ def abi_sum(toSum: abi.DynamicArray[abi.Uint64], *, output: abi.Uint64) -> Expr:
     def __init__(
         self,
         fn_implementation: Callable[..., Expr],
+        /,
+        *,
+        overriding_name: Optional[str] = None,
     ) -> None:
         self.output_kwarg_info: Optional[OutputKwArgInfo] = self._get_output_kwarg_info(
             fn_implementation
         )
         self.subroutine = SubroutineDefinition(
             fn_implementation,
             return_type=TealType.none,
+            name_str=overriding_name,
             has_abi_output=self.output_kwarg_info is not None,
         )
 
+    @staticmethod
+    def name_override(name: str) -> Callable[..., "ABIReturnSubroutine"]:
```

Comment:
Say I want a method to handle deposits,

`deposit_asa(axfer: abi.AssetTransferTransaction):`
`deposit_algo(ptxn: abi.PaymentTransaction):`

these can both technically have same same name `deposit` and different args (axfer vs payment) so the contract has no trouble routing them since the method selector is based on the hash of the full signature.

It provides an arguably nicer API to call the `deposit` method in either case with one of the 2 arguments but other issues arise like trying to get the method by name since they'll have the same name.

Reply:
I did see some reply by Ben under this thread, is that gone?

Reply:
@ahangsu Tagging you here because I see the method came in via a child PR - Can you add some context?

My primary concern is that any new public method raises a maintenance concern.  So, I'm sanity checking that there's usage in mind.  If not, I lean towards removing here.

Reply:
> is that gone?

Seems to be - I saw the same thing on refresh.  I just had a verbal with @barnjamin where we realized the reply was focused on ABI methods rather than subroutines.

Reply:
thanks folks, no worries on my side

Reply:
The context is that: considering `Subroutine` with overriding name, and in beaker: https://github.com/algorand-devrel/beaker/pull/84, folks want to override name of a method such that they can change something from:

```python
deposit_asa(axfer: abi.AssetTransferTransaction):
deposit_algo(ptxn: abi.PaymentTransaction):
```

to

```python
deposit(axfer: abi.AssetTransferTransaction):
deposit(ptxn: abi.PaymentTransaction):
```

The overriding name goes into multiple methods in `ABIReturnSubroutine`:
- method_selector (which goes into TEAL code generate)
- method_spec (which goes into `contract.json`)

We don't think it is fine to let method ask for such overriding name in multiple places, and such overriding name should get into initialization, such that the risk of misusing them can be minimized.

Without breaking the released API on `ABIReturnSubroutine`, we decide to introduce `override_name` method as a decorator method that initialize an `ABIReturnSubroutine`, and specialize its overriding name.

Reply:
seems resolved by #556